src/seq_mv/csr_spgemm_device_rocsparse.c

```diff
@@ -0,0 +1,180 @@
+
+   /* Copy the unsorted over as the initial "sorted" */
+   hypre_TMemcpy(d_ja_sorted, d_ja, HYPRE_Int,     nnzA, HYPRE_MEMORY_DEVICE, HYPRE_MEMORY_DEVICE);
+   hypre_TMemcpy(d_a_sorted,  d_a,  HYPRE_Complex, nnzA, HYPRE_MEMORY_DEVICE, HYPRE_MEMORY_DEVICE);
+   hypre_TMemcpy(d_jb_sorted, d_jb, HYPRE_Int,     nnzB, HYPRE_MEMORY_DEVICE, HYPRE_MEMORY_DEVICE);
+   hypre_TMemcpy(d_b_sorted,  d_b,  HYPRE_Complex, nnzB, HYPRE_MEMORY_DEVICE, HYPRE_MEMORY_DEVICE);
+
+   /* Sort each of the CSR matrices */
+   hypre_SortCSRRocsparse(m, k, nnzA, d_ia, d_ja_sorted, d_a_sorted);
+   hypre_SortCSRRocsparse(k, n, nnzB, d_ib, d_jb_sorted, d_b_sorted);
+
+   // nnzTotalDevHostPtr points to host memory
+   HYPRE_Int *nnzTotalDevHostPtr = &nnzC;
+   HYPRE_ROCSPARSE_CALL( rocsparse_set_pointer_mode(handle, rocsparse_pointer_mode_host) );
+
+   d_ic = hypre_TAlloc(HYPRE_Int, m+1, HYPRE_MEMORY_DEVICE);
+
+   // For rocsparse, we need an extra buffer for computing the
+   // csrgemmnnz and the csrgemm
+   //
+   // Once the buffer is allocated, we can use the same allocated
+   // buffer for both the csrgemm_nnz and csrgemm
+   //
+   // Note that rocsparse csrgemms do: C = \alpha*A*B +\beta*D
+   // So we hardcode \alpha=1, D to nothing, and pass NULL for beta
+   // to indicate \beta = 0 to match the cusparse behavior.
+   HYPRE_Complex alpha = 1.0;
+
+   size_t rs_buffer_size = 0;
+   void *rs_buffer;
+
+   if (isDoublePrecision)
+   {
+      HYPRE_ROCSPARSE_CALL( rocsparse_dcsrgemm_buffer_size(handle,
+                                                           transA, transB,
+                                                           m, n, k,
+                                                           &alpha, // \alpha = 1
+                                                           descrA, nnzA, d_ia, d_ja_sorted,
+                                                           descrB, nnzB, d_ib, d_jb_sorted,
+                                                           NULL, // \beta = 0
+                                                           NULL,   0,    NULL, NULL, // D is nothing
+                                                           infoC, &rs_buffer_size) );
+   }
+   else if (isSinglePrecision)
+   {
+      HYPRE_ROCSPARSE_CALL( rocsparse_scsrgemm_buffer_size(handle, transA, transB,
+                                                           m, n, k,
+                                                           (float *) &alpha, // \alpha = 1
+                                                           descrA, nnzA, d_ia, d_ja_sorted,
+                                                           descrB, nnzB, d_ib, d_jb_sorted,
+                                                           NULL, // \beta = 0
+                                                           NULL,   0,    NULL, NULL,
+                                                           infoC, &rs_buffer_size) );
+   }
+
+   HYPRE_HIP_CALL(hipMalloc(&rs_buffer, rs_buffer_size));
```

Comment:
Why not `hypre_TAlloc`?

Reply:
In this case, it was because this allocation didn't match the pattern of hypre_TAlloc (a count and a type). I wasn't sure if the preference would be to invoke hypre_MAlloc or just use the "native" API. Please let me know your preference and I'll update accordingly.

Reply:
I would prefer to wrap this in a hypre call, and `hypre_MAlloc` seems appropriate.  The "memory tracker" stuff would not be available for these calls, but that looks like a debugging aid, so maybe not so important.

@liruipeng .  What do you think?  Also, why do we have `hypre_MAlloc` and `_hypre_MAlloc'?

Reply:
> In this case, it was because this allocation didn't match the pattern of hypre_TAlloc (a count and a type). I wasn't sure if the preference would be to invoke hypre_MAlloc or just use the "native" API. Please let me know your preference and I'll update accordingly.

Not sure if I understand. Does this work, `rs_buffer = hypre_TAlloc(char, rs_buffer_size, HYPRE_MEMORY_DEVICE);`? I think we use this in many places where don't necessarily have a type but just want to allocate certain number of bytes. Thoughts?

Reply:
This is fine with me.  Does the compiler complain about type mismatches?

Reply:
> I would prefer to wrap this in a hypre call, and `hypre_MAlloc` seems appropriate. The "memory tracker" stuff would not be available for these calls, but that looks like a debugging aid, so maybe not so important.
> 
> @liruipeng . What do you think? Also, why do we have `hypre_MAlloc` and `_hypre_MAlloc`?

We prefer having all allocations via `hypre_TAlloc` so it goes through our memory model and can use memory pool potentially. `hypre_MAlloc` will also do, but as @rfalgout pointed out, it will not tracked by the "tracker", which is for debugging only. But I don't see where `hypre_MAlloc` works but `hypre_TAlloc` doesn't. 

`hypre_MAlloc` and `_hypre_MAlloc` were designed for different levels of memory space, conceptual for the former one and physical for the latter. In a few cases, we want to bypass the conceptual memory location but allocate on actual memory location, UVM vs device memory related, where we call `_hypre_MAlloc`. We also have the macro version of it, `_hypre_TAlloc`.

```
hypre_MAlloc(size_t size, HYPRE_MemoryLocation location)
_hypre_MAlloc(size_t size, hypre_MemoryLocation location)
```

Reply:
> This is fine with me. Does the compiler complain about type mismatches?

Good point. It may not since `rs_buffer` is `void *`. If it does, a pointer casting should work I think.

Reply:
Okay, thanks for the explanation @liruipeng .

Reply:
Hi @liruipeng you were correct, `rs_buffer = hypre_TAlloc(char, rs_buffer_size, HYPRE_MEMORY_DEVICE);` worked just fine. I have pushed that change in 6169f41. I've also grepped and made sure `hipMalloc` and `hipFree` only appear in `hypre_memory.c`. Sorry for the confusion on my part.